### ISSUE TITLE
Setting node outputs: nil pointer dereference in v2.4.1

### ISSUE THREAD

**Issue by jessesuen** [argoproj/argo#1691]
```
time="2019-10-17T13:51:22Z" level=info msg="Setting node magento-veke-ci-9s859[1].build (magento-veke-ci-9s859-2877268657) outputs"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x134ce7b]

goroutine 358 [running]:
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).addParamToGlobalScope(0xc0007990e0, 0xc000f2a778, 0x7, 0x0, 0x0, 0xc000a714c0, 0xc000f2a7a0, 0x7)
    /go/src/github.com/argoproj/argo/workflow/controller/operator.go:1712 +0x1cb
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).addOutputsToScope(0xc0007990e0, 0x16d5173, 0x8, 0xc000a71480, 0x0)
    /go/src/github.com/argoproj/argo/workflow/controller/operator.go:1610 +0x281
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).podReconciliation.func1(0xc000d77180)
    /go/src/github.com/argoproj/argo/workflow/controller/operator.go:599 +0x7a7
github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).podReconciliation.func2(0xc000f2a650, 0xc000e49e00, 0xc0007990e0, 0xc0008bc0e0, 0xc000817620, 0xc0003b3920, 0x0, 0x0, 0x0, 0x0, ...)
    /go/src/github.com/argoproj/argo/workflow/controller/operator.go:625 +0xaf
created by github.com/argoproj/argo/workflow/controller.(*wfOperationCtx).podReconciliation
    /go/src/github.com/argoproj/argo/workflow/controller/operator.go:623 +0x350
```

**Comment by paveq** [comment]
In our workflow we read a version file after CI process has completed:

```
    outputs:
      parameters:
      - name: version
        valueFrom:
          path: /tmp/version
        globalName: version
```

This error seemed to happen when main container exited before writing contents to /tmp/version, due to code not compiling properly. The issue is made worse as workflow gets stuck in running state, causing workflow controller to CrashLoopBackOff.

Logs from wait container:

```
time="2019-10-18T12:28:35Z" level=info msg="Main container completed"
time="2019-10-18T12:28:35Z" level=info msg="Saving output parameters"
time="2019-10-18T12:28:35Z" level=info msg="Saving path output parameter: version"
time="2019-10-18T12:28:35Z" level=info msg="Copying /tmp/version from base image layer"
time="2019-10-18T12:28:35Z" level=info msg="Annotations monitor stopped"
time="2019-10-18T12:28:35Z" level=info msg="[sh -c docker cp -a 248cfcb3d906443525453535bb0e4a2bed4b0d11d5a9b33f863173f97c1da226:/tmp/version - | tar -ax -O]"
time="2019-10-18T12:28:36Z" level=error msg="`[sh -c docker cp -a 248cfcb3d906443525453535bb0e4a2bed4b0d11d5a9b33f863173f97c1da226:/tmp/version - | tar -ax -O]` stderr:\nError: No such container:path: 248cfcb3d906443525453535bb0e4a2bed4b0d11d5a9b33f863173f97c1da226:/tmp/version\ntar: This does not look like a tar archive\ntar: Exiting with failure status due to previous errors\n"
time="2019-10-18T12:28:36Z" level=error msg="executor error: exit status 2\ngithub.com/argoproj/argo/errors.Wrap\n\t/go/src/github.com/argoproj/argo/errors/errors.go:88\ngithub.com/argoproj/argo/errors.InternalWrapError\n\t/go/src/github.com/argoproj/argo/errors/errors.go:71\ngithub.com/argoproj/argo/workflow/executor/docker.(*DockerExecutor).GetFileContents\n\t/go/src/github.com/argoproj/argo/workflow/executor/docker/docker.go:42\ngithub.com/argoproj/argo/workflow/executor.(*WorkflowExecutor).SaveParameters\n\t/go/src/github.com/argoproj/argo/workflow/executor/executor.go:430\ngithub.com/argoproj/argo/cmd/argoexec/commands.waitContainer\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/wait.go:44\ngithub.com/argoproj/argo/cmd/argoexec/commands.NewWaitCommand.func1\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/wait.go:16\ngithub.com/spf13/cobra.(*Command).execute\n\t/go/src/github.com/spf13/cobra/command.go:766\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/src/github.com/spf13/cobra/command.go:852\ngithub.com/spf13/cobra.(*Command).Execute\n\t/go/src/github.com/spf13/cobra/command.go:800\nmain.main\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/main.go:17\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:201\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1333"
time="2019-10-18T12:28:36Z" level=info msg="No output artifacts"
time="2019-10-18T12:28:36Z" level=info msg="Killing sidecars"
time="2019-10-18T12:28:36Z" level=info msg="Killing sidecar dind (d8240a7dfad7e20c3e589755a9f4ab7d0c049de4c7351f0b78252eb8b532766b)"
time="2019-10-18T12:28:36Z" level=info msg="docker kill --signal TERM d8240a7dfad7e20c3e589755a9f4ab7d0c049de4c7351f0b78252eb8b532766b"
time="2019-10-18T12:28:36Z" level=info msg="[docker wait d8240a7dfad7e20c3e589755a9f4ab7d0c049de4c7351f0b78252eb8b532766b]"
time="2019-10-18T12:28:36Z" level=info msg="Deadline monitor stopped"
time="2019-10-18T12:28:37Z" level=info msg="Containers [d8240a7dfad7e20c3e589755a9f4ab7d0c049de4c7351f0b78252eb8b532766b] killed successfully"
time="2019-10-18T12:28:37Z" level=info msg="No Script output reference in workflow. Capturing script output ignored"
time="2019-10-18T12:28:37Z" level=info msg="Annotating pod with output"
time="2019-10-18T12:28:37Z" level=info msg="Alloc=5677 TotalAlloc=12383 Sys=70590 NumGC=4 Goroutines=8"
```

**Comment by jessesuen** [comment]
We reproduced this in v2.4.1 and confirmed this is fixed in the release-2.4 branch through some of the template resolution refactoring. Will be included in a v2.4.2 patch next week
